docs/Overview.md

```diff
@@ -1,96 +1,246 @@
-    ID() NodeID
-    // More details about what the node does
+type Result interface {
+    Name() string
+    // ... methods to iterate other the contents of the result
 }
+```
 
-// NodeID uniquely identifies a node.
-type NodeID
+The executor executes a query expressed in some language and fills its results, provided an execution context.
+There can be multiple executors, one for each language supported by the engine.
+The engine is responsible for providing the executor with an
+execution context that embeds everything that is necessary for the execution of the query, included the registered values,
+sources and transformations. 
 
-// Edge establishes a parent child relationship between two nodes.
-type Edge interface {
-    Parent() NodeID
-    Child() NodeID
+```go
+type Executor interface {
+    Execute(Query, ExecutionContext) error // fills query results.
+    Type() Language
 }
 
-// Planner computes a plan from a query and available storage interfaces
-type Planner interface {
-    Plan(Query, []Storage) Plan
+// The ExecutionContext contains any information necessary for properly executing a query.
+type ExecutionContext interface {
+    Bindings() Scope
+    GetSource(id string) Source
+    GetTransformation(id string) Transformation
+    
+    // ... examples of other components of the execution context.
+    Context() context.Context
+    MemoryAllocator() Allocator
+    Logger() Logger
 }
 
-// Plan is a DAG of the specific steps to execute.
-type Plan interface  {
-    Nodes() []PlanNode
-    Edges() []Edge
+// Scope is a set of bindings from identifiers to values.
+type Scope = map[string]Value
+```
+   
+A compiler translates a query representation in a language into another target language.
+
+```go
+type Compiler interface {
+    Compile(Representation) (Representation, err)
+    Type() CompilerType
 }
+```
 
-// PlanNode is a single step in the plan DAG.
-type PlanNode interface {
-    ID() NodeID
-    Predicates() []Predicate
+As an example, let `e` be an engine that supports `String`, `AST`, and `SemanticGraph` queries:
+
+```go
+func (e *Engine) Query(r Representation) (Query, error) {
+    var e Executor
+    switch l := r.Language() {
+        case "String":
+            e = &StringExecutor{...}
+        case "AST":
+            e = &ASTExecutor{...}
+        case "SemanticGraph":
+            // the interpreter is an Executor!
+            e = interpreter.NewInterpreter()
+        default:
+            return nil, fmt.Errorf("unknown language: %v", l)
+    }
+    
+    q := ... // new query
+    ec := ... // create execution context
+    err := e.Execute(q, ec)
+    return q, err
 }
+```
 
-// Predicate filters data.
-type Predicate interface {}
+The `StringExecutor` does not actually know how to execute a script, but it uses a compiler to translate the script into its
+AST representation and delegate to an `ASTExecutor` in turn:
+
+```go
+func (e *StringExecutor) Execute(q Query, ec ExecutionContext) error {
```

Comment:
I think this works badly because it tightly couples the compiler and the executor. I think separating them makes it possible for us to translate between different representations and execute them separately. If we do this, then we tightly couple them and there is only one way to execute an AST. This makes it impossible to do things like proxies.

Reply:
We absolutely need to talk more about this. I am not against my design (obviously) neither against yours. Indeed, I know we talked about compilation and execution as 2 separated phases.

However, when designing  this, I understood that I couldn't force an executor not to compile to another representation and delegate to another executor. I think that I need to talk to better understand this:

> If we do this, then we tightly couple them and there is only one way to execute an AST.

The `ASTExecutor` knows the path to reach another executor which it can delegate execution. I understand that they are coupled now, but, how can you deal with having multiple executors for the same language? How should the engine decide on which to use?

An alternative approach could be:
 - register compilers and executors through `Create*` functions.
 - the engine would be the only one able to instantiate executors and compilers.
 - it compiles until it finds an executable representation (if it doesn't find one, it could proxy to another engine that has it).
 - the executor cannot instantiate compilers neither other executors, the only thing it can do is to execute.

I don't know if this is possible:

> the executor cannot instantiate compilers neither other executors

Or should be enforced as a good practice.